lib/wireprotocol/3_2_support.js

```diff
@@ -1,19 +1,67 @@
 'use strict';
 
-var Query = require('../connection/commands').Query,
-  retrieveBSON = require('../connection/utils').retrieveBSON,
-  f = require('util').format,
-  MongoError = require('../error').MongoError,
-  MongoNetworkError = require('../error').MongoNetworkError,
-  getReadPreference = require('./shared').getReadPreference;
-
-var BSON = retrieveBSON(),
-  Long = BSON.Long;
+const Query = require('../connection/commands').Query;
+const retrieveBSON = require('../connection/utils').retrieveBSON;
+const f = require('util').format;
+const MongoError = require('../error').MongoError;
+const MongoNetworkError = require('../error').MongoNetworkError;
+const getReadPreference = require('./shared').getReadPreference;
+const incrementStatementId = require('../topologies/shared').incrementStatementId;
+const BSON = retrieveBSON();
+const Long = BSON.Long;
 
 var WireProtocol = function(legacyWireProtocol) {
   this.legacyWireProtocol = legacyWireProtocol;
 };
 
+/**
+ * Optionally decorate a command with transactions specific keys
+ *
+ * @param {Object} command the command to decorate
+ * @param {ClientSession} session the session tracking transaction state
+ */
+function decorateWithTransactionsData(command, session) {
+  if (!session) {
+    return;
+  }
+
+  // first apply non-transaction-specific sessions data
+  const serverSession = session.serverSession;
+  if (serverSession.txnNumber) {
+    command.txnNumber = BSON.Long.fromNumber(serverSession.txnNumber);
+  }
+
+  // now try to apply tansaction-specific data
+  if (!session.inTransaction()) {
+    return;
+  }
+
+  command.stmtId = serverSession.stmtId;
+  command.autocommit = false;
+
+  if (serverSession.stmtId === 0) {
+    command.startTransaction = true;
+    command.readConcern = { level: 'snapshot' };
+
+    if (session.transactionOptions.readConcern) {
```

Comment:
```js
const readConcern = session.transactionOptions.readConcern || session.clientOptions.readConcern;
if (readConcern) {
  command.readConcern = readConcern;
}
```

Also may need to guard against `transactionOptions` and/or `clientOptions` not being there.